.teamcity/test/AllProjectsTest.kt

```diff
@@ -31,18 +31,47 @@ class AllProjectsTest {
                 ProjectRelease to "ProjectKtorRelease"
         )
 
-        projectsFromPackage("subprojects").forEach { project ->
+        allProjects().forEach { project ->
             assertTrue(projectsIDs.containsKey(project), "Cannot find expected ID for project ${project.name}")
             assertEquals(projectsIDs[project], project.id.toString(), "ID for Project ${project.name} doesn't match")
             assertHasUniqueID(project)
         }
     }
 
+    @Test
+    fun projectsBuildTypesHadDefinedIDs() {
+        val buildTypesIDs = mapOf(
+                ProjectCore to setOf(
+                        "KtorMatrixMacOSJava8",
+                        "KtorMatrixMacOSJava11",
+                        "KtorMatrixLinuxJava8",
+                        "KtorMatrixLinuxJava11",
+                        "KtorMatrixWindowsJava8",
+                        "KtorMatrixWindowsJava11"
+                ),
+                ProjectDocSamples to setOf("KtorDocs_ValidateSamples")
+        )
+
+        allProjects().forEach { project ->
+            if (project.buildTypes.size > 0) {
+                assertTrue(buildTypesIDs.containsKey(project), "Cannot find build types IDs for project ${project.name}")
+                val ids = project.buildTypes.map { it.id.toString() }.toSet()
+                assertEquals(buildTypesIDs[project], ids, "Build types IDs for project ${project.name} doesn't match")
+            }
+        }
+    }
+
     private fun assertHasUniqueID(project: Project) {
         val className = project.javaClass.name.split('.').last()
         assertNotEquals(className, project.id.toString(), "Project ID '${project.id.toString()}' should differ from its class name")
     }
 
+    private fun allProjects(): List<Project> {
+        val projects = projectsFromPackage("subprojects")
+        assertTrue(projects.isNotEmpty(), "No projects found")
```

Comment:
Is there a reason there is an assertion in the private method? I'm concerned about having many functions and logic in tests, where one would need to end up testing the actual logic of helper methods. If the methods are very small and understandable then it could be acceptable. However, I do think assertions should remain as part of the test itself and not in any helper methods.